### PR TITLE
Transport spec diagrams for #54

### PR DESCRIPTION
**Pull request by MikeBishop**
Updates the transport spec diagrams to the format @larseggert requested.  A few related changes along the way that are hopefully uncontroversial:

- Added a notation conventions section, à la RFC 5201, describing the things we're adding to RFC 2360 (including the ones borrowed from 5201).
- Shifted to using the term "payload" rather than "headers" and "body," since those are never actually used in the descriptions of subsequent packets or frames.  (Describing frame payloads rather than repeating the Type each time simplifies the diagrams and makes them align more nicely.)
- Totally removed the Diversification Nonce field, since that's optional and dependent on a removed bit (in [32e9d5](https://github.com/quicwg/base-drafts/commit/abf5b35a047f33e75f5d31a2b65c26699532e9d5))
- Removed most of the text and the diagram from the Public Reset, since that depends on specs not in evidence (QUIC-Crypto) and I didn't want to do a diagram of an obsolete packet format.  We'll augment EKR's PR with a diagram later.
- Converted the figure-of-a-table of frame types to an actual table, with references to each section.

### CODE REVIEW

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -943,47 +940,104 @@ The fields in the ACK frame are as follows:
...
 
-  * First Timestamp (opt): An optional 32-bit unsigned value specifying the time
-    delta in microseconds, from the beginning of the connection to the arrival
-    of this packet.
+* First Timestamp (opt): An optional 32-bit unsigned value specifying the time
+  delta in microseconds, from the beginning of the connection to the arrival
+  of this packet.
 
-  * Delta Largest Observed (opt, repeated): (Same as above.)
+* Delta Largest Acked 1-N (opt, repeated): (Same as above.)
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: Maybe use an ellipsis here rather than the hyphen, which my addled brain interpreted as minus.

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -1028,24 +1082,27 @@ The fields in the STOP_WAITING frame are as follows:
...
-The WINDOW_UPDATE frame informs the peer of an increase in an endpoint's flow
-control receive window.  The StreamID can be zero, indicating this WINDOW_UPDATE
-applies to the connection level flow control window, or non-zero, indicating
-that the specified stream should increase its flow control window.  The frame is
-as follows:
+The WINDOW_UPDATE frame (type=0x04) informs the peer of an increase in an 
+endpoint's flow control receive window. The Stream ID can be zero, indicating 
+this WINDOW_UPDATE applies to the connection level flow control window, or 
+non-zero, indicating that the specified stream should increase its flow control 
+window. The frame is as follows:
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: You need to watch for trailing whitespace here.

File: draft-ietf-quic-transport.md [hunk 3]
```diff
@@ -1103,80 +1164,70 @@ The fields are:
...
-~~~
+The PADDING frame pads a packet with 0x00 bytes. When this frame is encountered, 
+the rest of the packet is expected to be padding bytes. The frame contains 0x00 
+bytes and extends to the end of the QUIC packet. A PADDING frame has no payload, 
+and must have the 8-bit Frame Type field set to 0x00. 
 
-## PING frame
+## PING frame {#frame-ping}
 
 Endpoints can use PING frames to verify that their peers are still alive or to
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: (type=0x...)

File: draft-ietf-quic-transport.md [hunk 4]
```diff
@@ -1000,24 +1054,24 @@ actual exponent is one-less than the explicit exponent, and the value represents
...
+(specified in the QUIC Frame packet's Flags field.) The frame is as follows: 
 
 ~~~
-   +---------------------------------------------------+
-   | Type (8) | Least unacked delta (8, 16, 32, or 48) |
-   +---------------------------------------------------+
+ 0                   1                   2                   3
+ 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|               Least Unacked Delta (8/16/32/24)              ...
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: This is not a faithful transcription.  s/24/48

File: draft-ietf-quic-transport.md [hunk 5]
```diff
@@ -404,33 +419,32 @@ The fields in a Regular packet past the Common Header are the following:
...
-  nonce should be removed from the packet format.
-
 * Packet Number: The lower 8, 16, 32, or 48 bits of the packet number, based on
   the PACKET_NUMBER_SIZE flag.  Each Regular packet is assigned a packet number
   by the sender.  The first packet sent by an endpoint MUST have a packet number
   of 1.
 
-* AEAD Data: A Regular packet's header, which includes the Common Header, and
-  the Version, Diversification Nonce, and Packet Number fields, is authenticated
+* AEAD Data: A Regular packet's Common Header and fields are authenticated,
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: I don't think that it's accurate to say that the Regular packet's Common Header and fields are authenticated, since the "fields" contain frames and those are encrypted. Perhaps we should start calling them Regular packet header and payload, so that the Common Header and the Regular packet's Header are authenticated, and the Regular packet's payload is encrypted?

Comment (martinthomson) [inline draft-ietf-quic-transport.md]: I think that we can merge this absent any fix here.  It's a minor point that needs to be corrected, but this PR doesn't need to fix problems it didn't create :)

File: draft-ietf-quic-transport.md [hunk 6]
```diff
@@ -374,27 +398,18 @@ Check the flags in the common header
 
 ## Regular Packets
 
-Each Regular packet's header consists of a Common Header followed by fields
-specific to Regular packets, as shown below:
+Each Regular packet's payload consists of fields shown below:
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: I would change this to: "Each Regular packet contains additional header fields followed by payload, as shown below:"

File: draft-ietf-quic-transport.md (hunk 5 above)
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: In addition to the suggested change above, I would change this to: "A Regular packet's header fields, including the Common Header, are authenticated but not encrypted."

Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: This PR introduces it. The text currently is quite clear about what's authenticated and what's encrypted. This new text reads better, but confuses that point. I've suggested some mods.

File: draft-ietf-quic-transport.md [hunk 7]
```diff
@@ -805,46 +805,50 @@ describe the various QUIC frame types that can be present in a Regular packet.
...
+|                      [Data Length (16)]                       |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                        Stream Data (*)                      ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+~~~
+{: #stream-format title="STREAM Frame Format"}
 
-  * The 'SS' bits encode the length of the Stream ID header field as 8, 16, 24,
-    or 32 bits.  (DISCUSS: Consider making this 8, 16, 32, 64.)
+The STREAM frame payload contains the following fields:
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: I don't think it makes sense to call this the payload. Why not just say "In addition to the Type byte, the STREAM frame contains the following fields"?

File: draft-ietf-quic-transport.md [hunk 8]
```diff
@@ -943,47 +940,104 @@ The fields in the ACK frame are as follows:
 * Ack Delay: Time from when the largest acked, as indicated in the Largest Acked
   field, was received by this peer to when this ack was sent.
 
-* Ack Block Section:
+* Ack Block Section: Contains one or more blocks of packet numbers which have
+  been successfully received.  See {{ack-block-section}}.
+  
+* Timestamp Section: Contains zero or more timestamps reporting transit delay of
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: "transit delay" --> "receive times"

File: draft-ietf-quic-transport.md [hunk 9]
```diff
@@ -943,47 +940,104 @@ The fields in the ACK frame are as follows:
...
 
-  * Ack Block Length (opt, repeated): An unsigned packet number delta that
-    indicates the number of contiguous packets being acked starting after the
-    end of the previous gap.  Along with the previous field, this field is
-    repeated "Num Blocks" times.
+### Timestamp Section {#timestamp-section}
+
+The Timestamp Section contains between zero and 255 measurements of packet
+delay. This information can be used by a sender to better estimate the RTT of
+the connection.
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: The timestamps are simply packet receive times as observed by the receiver, and are not packet transit times. How about something like: "... of packet receive time deltas. This information may be used by a sender to better estimate the RTT of the connection."

File: draft-ietf-quic-transport.md [hunk 10]
```diff
@@ -1055,46 +1112,50 @@ The fields in the WINDOW_UPDATE frame are as follows:
...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                        Stream ID (32)                         |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 ~~~
 
-The fields in the BLOCKED frame are as follows:
-
-* Frame Type: The Frame Type byte is an 8-bit value that must be set to 0x05
-  indicating that this is a BLOCKED frame.
+The BLOCKED frame payload contains a single field:
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: I would avoid saying "payload" here -- the BLOCKED frame has no payload really.

File: draft-ietf-quic-transport.md [hunk 11]
```diff
@@ -1103,80 +1164,70 @@ The fields are:
...
-only has a Frame Type field, and must have the 8-bit Frame Type field set to
-0x00. The PADDING frame is as follows:
+## PADDING Frame {#frame-padding}
 
-~~~
-   +--------+
-   |  0x00  |
-   +--------+
-~~~
+The PADDING frame pads a packet with 0x00 bytes. When this frame is encountered,
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: For parity with other frames, perhaps add "(type = 0x00)" here

File: draft-ietf-quic-transport.md (hunk 5 above)
Comment (MikeBishop) [inline draft-ietf-quic-transport.md]: Well, one of the (authenticated) fields contains ciphertext, which then contains frames....  I think that's a separate terminology issue than Martin's comment, isn't it?

File: draft-ietf-quic-transport.md (hunk 7 above)
Comment (MikeBishop) [inline draft-ietf-quic-transport.md]: We can phrase it either way, but I was trying to get a consistent usage that for both packets and frames, there's a type-independent envelope (which for frames is really thin) and a type-dependent payload.

File: draft-ietf-quic-transport.md (hunk 10 above)
Comment (MikeBishop) [inline draft-ietf-quic-transport.md]: Same comment -- let's agree on terminology first.  Per the description at line 490, this is a payload.  If we want a different term, let's sync on that first.

Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: Now that I understand what you were going for, I think the use of "payload" is perfectly accurate, but adds an unnecessary level of abstraction: we'll end up referring to "fields in the BLOCKED frame" anyways. In addition, we lose the ability to use "payload" for stream data that's carried within a STREAM frame. I'd prefer that we simply called it the frame's fields instead of the frame's payload.

File: draft-ietf-quic-transport.md [hunk 12]
```diff
@@ -943,47 +940,104 @@ The fields in the ACK frame are as follows:
...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|  [Gap 1 (8)]  |       [Ack Block 1 Length (8/16/32/48)]     ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|  [Gap 2 (8)]  |       [Ack Block 2 Length (8/16/32/48)]     ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+                             ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|  [Gap N (8)]  |       [Ack Block N Length (8/16/32/48)]     ...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+~~~
```
Comment (MikeBishop) [inline draft-ietf-quic-transport.md]: Per Jana:  Move numbers from each section into the fixed structure above.

File: draft-ietf-quic-transport.md [hunk 13]
```diff
@@ -943,47 +940,104 @@ The fields in the ACK frame are as follows:
...
+* Delta Largest Acked (opt): An optional 8-bit unsigned packet number delta
+  specifying the delta between the largest acked and the first packet whose
+  timestamp is being reported.  In other words, this first packet number may
+  be computed as (Largest Acked - Delta Largest Acked.)
 
-  * Time Since Previous Timestamp (opt, repeated): An optional 16-bit unsigned
-    value specifying time delta from the previous reported timestamp.  It is
-    encoded in the same format as the Ack Delay.  Along with the previous field,
-    this field is repeated "Num Timestamps" times.
+* First Timestamp (opt): An optional 32-bit unsigned value specifying the time
```
Comment (MikeBishop) [inline draft-ietf-quic-transport.md]: Inconsistent use - First Timestamp vs. Time Since Largest Acked.  Per discussion with Jana, this doesn't need to still be 32 bits; @janaiyengar will file issue to trim it down.

File: draft-ietf-quic-transport.md (hunk 5 above)
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: > A Regular packet's header fields, including the Common Header, are authenticated but not encrypted.

That would be incorrect.

Comment (martinthomson) [inline draft-ietf-quic-transport.md]: I should expand on that a little:

Regular (ugh, that name has to change) packets aren't always protected.  When they are, the header fields are not protected at all.  However, I believe that the values that are present do not require protection.  Outside of 0-RTT, the only fields that need to be protected are the flags field, the connection ID and the packet number.

Connection ID is easy not to authenticate, it just tells you which key to use.  Change it and you are guaranteed (or at least with astonishingly high probability) to get the wrong keys.

Packet number is input to the nonce, so that's covered.

Of the flags, the only relevant one is KEY_PHASE, which tells the receiver which key to use (see connection ID).  The packet number size bits are subsumed by the packet number.  I have no idea what the multipath bit might do.  Public reset and version don't appear on regular packets with one exception.  Version does appear on 0-RTT packets, but it serves the same purpose as KEY_PHASE there.

File: draft-ietf-quic-transport.md (hunk 12 above)
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: I wouldn't do this just yet. I'd leave this PR to not do major changes to the ACK format, and leave that for a separate PR that reworks the ack format entirely.

File: draft-ietf-quic-transport.md (hunk 10 above)
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: That makes sense to me.  Then we can talk about "transport protocol payload" and "STREAM frame payload" and we will be talking about the same thing.

File: draft-ietf-quic-transport.md (hunk 12 above)
Comment (MikeBishop) [inline draft-ietf-quic-transport.md]: @janaiyengar, I think your comment is for a different change. My other note said you'd file an issue for reducing the size of the `Time Since Largest Acked` field to match the others, and I agree that should be saved for a different PR.  This is a purely editorial change which you suggested, putting the counts up in the top section.  If you no longer want that, we can revert it.

Comment (MikeBishop) [inline draft-ietf-quic-transport.md]: ...but I also had a note that I had an inconsistency between `Time Since Largest Acked` and `First Timestamp`.  The description of First Timestamp gives it as a time since the beginning of the connection, which makes more sense to be larger, and the remainders are deltas from that value.

**Comment by MikeBishop** [comment]
At this point, I think I've incorporated Jana's feedback and it addresses #54.  @janaiyengar, @martinthomson, @larseggert, please confirm.

**Comment by martinthomson** [comment]
@janaiyengar, I'd really like to merge this.  Is there any reason I shouldn't?

File: draft-ietf-quic-transport.md [hunk 14]
```diff
@@ -947,47 +947,104 @@ The fields in the ACK frame are as follows:
...
+* Ack Block Length (opt, repeated): An unsigned packet number delta that
+  indicates the number of contiguous packets being acked starting after the
+  end of the previous gap.  Along with the previous field, this field is
+  repeated "Num Blocks" times.
+
+### Timestamp Section {#timestamp-section}
+
+The Timestamp Section contains between zero and 255 measurements of packet
+receive times. This information can be used by a sender to better estimate the
+RTT of the connection.
```
Comment (ianswett) [inline draft-ietf-quic-transport.md]: I would leave the second sentence out, because I don't think improving RTT estimates are the most likely use case.

I would suggest adding "relative to the beginning of the connection." to the end of the first sentence to clarify these aren't an effort to convey a true time.

**Comment by janaiyengar** [comment]
LGTM -- looks great, ship it!
